### PR TITLE
Some Changes to NDArray Interface

### PR DESCRIPTION
**Pull request by ZihengJiang**
- Remove `NDArray::raw_data`
- Change `NDArray::offset_` to `byte_offset_`

@tqchen @piiswrong

### CODE REVIEW

File: include/mxnet/tensor_blob.h [hunk 1]
```diff
@@ -306,6 +288,16 @@ class TBlob {
         this->shape_.FlatTo3D(axis_begin, axis_end), stream);
   }
 
+  inline void SetDLTensor(int dev_mask, int dev_id) {
```
Comment (ZihengJiang) [inline include/mxnet/tensor_blob.h]: I'm not sure whether we should make this public or let NDArray access this by friend method

Comment (tqchen) [inline include/mxnet/tensor_blob.h]: friend method is better

**Comment by tqchen** [comment]
LGTM

File: include/mxnet/ndarray.h [hunk 2]
```diff
@@ -421,15 +394,25 @@ class NDArray {
...
   std::shared_ptr<MKLMemHolder> Mkl_mem_;
 #endif
   /*! \brief internal data of NDArray */
   std::shared_ptr<Chunk> ptr_;
   /*! \brief shape of current NDArray */
   TShape shape_;
-  /*! \brief offset in chunk */
-  size_t offset_;
+  /*! \brief byte offset in chunk */
+  size_t byte_offset_ = 0;
```
Comment (piiswrong) [inline include/mxnet/ndarray.h]: why?

Comment (tqchen) [inline include/mxnet/ndarray.h]: mainly to avoid additional switch on dtype during calculation, this is in bytes

Comment (piiswrong) [inline include/mxnet/ndarray.h]: but then you need to switch when converting back, like in ndarray.cc

Comment (tqchen) [inline include/mxnet/ndarray.h]: there is no need to switch, the only problem is slice, which is rare. However, there are more conventions that follows byte_offset_ like in DLPack. 

In the future, you should be able derive the byte cost from type(e.g. DLType), removing the need of switch.

**Comment by tqchen** [comment]
@piiswrong merge?

**Pull request merged**